File: GCTP_CPP/sinusoidal.cpp
#include "sinusoidal.h"

#include <stdio.h>
#include <math.h>


Sinusoidal::Sinusoidal(): Pseudocylindrical()
{
	setName("Sinusoidal");
	setNumber(SNSOID);
}

Sinusoidal::Sinusoidal( double gctpParameters[15], ProjUnit units, ProjDatum dat) 
: Pseudocylindrical( gctpParameters, units, dat )
{
  setName("Sinusoidal");
  setNumber(SNSOID);
}

void Sinusoidal::forward ( double lon, double lat, double* x, double* y )
{
  double deltaLon;	/* Delta longitude (Given longitude - center */
  
  clearError();

  if(m_initNeeded)
	init();

  //convert lat/lon from dec degrees to radians
  Util::convertCoords(DEGREE, RADIAN, lat, lon);

  /* Forward equations */
  deltaLon = Util::adjust_lon(lon - m_centerLon);

  m_x_coord = m_radius * deltaLon * cos( lat ) + m_falseEasting;
  m_y_coord = m_radius * lat + m_falseNorthing;

  Util::convertCoords(METER, m_unitCode, m_x_coord, m_y_coord);

  if( x != NULL )
  {
     *x = m_x_coord;
  }

  if( y != NULL )
  {
     *y = m_y_coord;
  }
}

void Sinusoidal::inverse ( double x, double y, double* lon, double* lat )
{
  double temp;		/* Re-used temporary variable */
  
  clearError();
  
  if(m_initNeeded)
	init();
  //convert lat/lon from dec degrees to radians
  Util::convertCoords(m_unitCode, METER, x, y);

  x -= m_falseEasting;
  y -= m_falseNorthing;

  m_latitude = y / m_radius;

  if( fabs( m_latitude ) > HALF_PI) {
		setError(164);
		return;
  }


  temp = fabs( m_latitude ) - HALF_PI;
  if( fabs( temp ) > EPSLN )
  {
     temp = m_centerLon + x / (m_radius * cos( m_latitude ));
	 m_longitude = Util::adjust_lon( temp );
  }
  else
  {
     m_longitude = m_centerLon;
  }

  Util::convertCoords(RADIAN, DEGREE, m_longitude, m_latitude);

  if( lon != NULL )
  {
     *lon = m_longitude;
  }

  if( lat != NULL )
  {
     *lat = m_latitude;
  }

}

void Sinusoidal::init(  )
{
  clearError();
  m_initNeeded = false;
}

File: JGCTP/sinusoidal.java
package JGCTP;

public class Sinusoidal extends Projection
{
	public Sinusoidal()
	{
		super();
		setNumber(ProjCode.SNSOID);
		setName("Sinusoidal");
	}
	
	public Sinusoidal(double gctpParams[], ProjUnit units, ProjDatum dat)
	{
		super(gctpParams, units, dat);
		setNumber(ProjCode.SNSOID);
		setName("Sinusoidal");
	}
	
	protected void _inverse(CoordPoint _p) throws ProjException
	{
	    double temp;		/* Re-used temporary variable */
	    CoordPoint p = new CoordPoint(_p);
	    
	    p.x -= m_falseEasting;
	    p.y -= m_falseNorthing;

	    m_lonLat.lat = p.y / m_sphere.radius;

	    if( Math.abs( m_lonLat.lat ) > Constants.HALF_PI) 
	    {
	  		throw(new ProjException(164, "Sinusoidal._inverse()"));
	    }

	    temp = Math.abs( m_lonLat.lat ) - Constants.HALF_PI;
	    if( Math.abs( temp ) > Constants.EPSLN )
	    {
	       temp = m_center.lon + p.x / (m_sphere.radius * Math.cos( m_lonLat.lat ));
	  	   m_lonLat.lon = Util.adjust_lon( temp );
	    }
	    else
	    {
	       m_lonLat.lon = m_center.lon;
	    }
	}
	
	protected void _forward(GeoPoint _p)
	{
	    double deltaLon;	/* Delta longitude (Given longitude - center */
	    
	    GeoPoint p = new GeoPoint(_p);
	    /* Forward equations */
	    deltaLon = Util.adjust_lon(p.lon - m_center.lon);

	    m_xy.x = m_sphere.radius * deltaLon * Math.cos(p.lat ) + m_falseEasting;
	    m_xy.y = m_sphere.radius * p.lat + m_falseNorthing;

	}
}